lib/fixed-points.nix

```
@@ -394,4 +394,64 @@ rec {
     fix' (self: (rattrs self) // {
       ${extenderName} = f: makeExtensibleWithCustomName extenderName (extends f rattrs);
     });
+
+  /**
+    Convert to an extending function (overlay).
+
+    `toExtension` is the `toFunction` for extending functions (a.k.a. extensions or overlays).
+    It converts a non-function or a single-argument function to an extending function,
+    while returning a double-argument function as-is.
+
+    That is, it takes one of `x`, `prev: x`, or `final: prev: x`,
+    and returns `final: prev: x`, where `x` is not a function.
+
+    This function is extracted from the implementation of
+    the fixed-point arguments support of `stdenv.mkDerivation`.
+    It bridges the gap between `<pkg>.overrideAttrs`
+    before and after the overlay-style support,
+    as well as `config.packageOverrides` and `config.overlays` in `pkgs`.
+
+    # Inputs
+
+    `f0`
+    : The function or non-function to convert to an extending function.
+
+    # Type
+
+    ```
+    toExtension ::: (a | a -> a | a -> a -> a) -> a -> a -> a
+    a = AttrSet & !Function
```

This is a bit more general and accurate.
- Implementation doesn't actually care about attribute sets, and would work equally well for lists, insofar as lists work well for anything.
- Make better use of parametric polymorphism. (`a` felt more like a "top" type, such as `Object` in Java or `unknown` in TypeScript)

```suggestion
    toExtension ::
      (data | b -> data | a -> b -> c) ->
        a -> b -> c
    where data = ! Callable
```

---

Thank you for reviewing!

> ```nix
> toExtension ::
>   (data | b -> data | a -> b -> c) ->
>     a -> b -> c
> where data = ! Callable
> ```

Why are we using `c` instead of `data` here?

---

Good point


> Why are we using `c` instead of `data` here?

- In the argument, in the latter `|` case, `c` could be anything, including a function.
- `c` can be returned by the returned extension function
  - but so can `data`!


```suggestion
    toExtension ::
      (data | b -> data | a -> b -> c) ->
        a -> b -> (data | c)
    where data = ! Callable
```

----

Do you think this is easier to read?

```suggestion
    toExtension ::
      ( data
      | b -> data
      | a -> b -> c
      ) ->
      a -> b -> (data | c)
    where data = ! Callable
```

---

I think it's reasonable to expect `c = data`, as our goal is to construct a function in the form `final: prev: dataTemplate`, and non-function data will be returned directly.

Is there a counter-example in which `c != data`?

---

There is. Here's an example that returns a function instead of attrs. I wouldn't recommend it, because the function needs to be invoked to be useful, which means these attributes aren't lazy like when you return them directly.

```nix
let
  extendsWith = op: overlay: f: final: let prev = f final; in op prev (overlay final prev)
in

fix
  (extendsWith
    (p: q: x: p x // q x)
    (toExtension (final: prev:
      x:
        let pre = prev x; fin = final x;
        in
          { a = 1; b = pre.a; c = fin.a + x; }))
    (final: x: { a = 0; c = final.a; }))
  42
```

---

I now realize that `data` might be too suggestive. An attribute set _containing_ a function would be fine.

---

The purpose of `toExtension` is to make `toExtension (final: prev: resultTemplate)`, `toExtension (prev: resultTemplate)`, and `toExtension result` all evaluate to `final: prev: resultTemplate`, which doesn't hold when the result is a function/callable.

Another implementation assumption is that `(final: prev: resultTemplate)` could take `prev` without error, which suggests that the "`prev`-space" is a subset of the "`final`-space." Therefore, the `prev` type is a sub-type of `final`. They are likely to share the same type in Nix Language.

---

What do you think about this?

```suggestion
    toExtension ::
      (a | a -> b | a -> a -> b) -> a -> a -> b
    where a = ! Callable
```

For example:
```nix
toExtension { a = 1; b = 2; c = null; }
=> final: prev: { a = 1; b = 2; c = null; }
```
```nix
toExtension (prev: { a = 1; b = elemAt prev 0; c = null; })
=> final: prev: { a = 1; b = elemAt prev 0;  c = null; }
```
```nix
toExtension (final: prev: { a = 1; b = elemAt prev 0; c = (elemAt final 0) + 1; })
=> final: prev: { a = 1; b = elemAt prev 0;  c = (elemAt final 0) + 1; }
```

We could replicate the documented examples with custom `fix'` and `extends'`:

```nix
elemAt = lib.elemAt

findFirst = lib.findFirst

toExtension = lib.toExtension

zipAttrsWith = lib.zipAttrsWith

recover = x: [ x.a x.b x.c ]

update' = lSet: rSet: zipAttrsWith (name: findFirst (x: x != null) null) [ rSet lSet ]

fix' = f: let x = f (recover x); in x

extends' = g: f: final: update' (f final) (g final (recover (f final)))

fix' (final: { a = 0; b = null; c = elemAt final 0; })
=> { a = 0; b = null; c = 0 }

fix' (extends' (toExtension { a = 1; b = 2; c = null; }) (final: { a = 0; b = null; c = elemAt final 0; }))
=> { a = 1; b = 2; c = 1; }

fix' (extends' (toExtension (prev: { a = 1; b = elemAt prev 0; c = null; })) (final: { a = 0; b = null; c = (elemAt final 0) + 1; }))
=> { a = 1; b = 0; c = 2; }
```

---

> The purpose of `toExtension` is to make `toExtension (final: prev: resultTemplate)`, `toExtension (prev: resultTemplate)`, and `toExtension result` all evaluate to `final: prev: resultTemplate`, which doesn't hold when the result is a function/callable.

This means that `toExtension` is subjectively not a great fit for such a use case, but it doesn't mean that it is incapable of producing a useful result. It can still be used for the only input type that isn't "misclassified", `a -> a -> b`.
It does not mean that the type of the function is less general than it really is.

I think when writing types, we should try to err on the side of being more general, so that we don't mislead users into thinking that a function is not applicable to their use case. There's no way to find or correct that mistake, so I'd rather have them try and fail. That might sound bad but I don't think it's a big deal, because I expect such false positives to be quite rare, especially because we also have documentation for them to judge the function's fitness for their purpose.


> Another implementation assumption is that `(final: prev: resultTemplate)` could take `prev` without error, which suggests that the "`prev`-space" is a subset of the "`final`-space." Therefore, the `prev` type is a sub-type of `final`. They are likely to share the same type in Nix Language.

This is not an intrinsic property of the function, so I don't think it belongs in the type.
For instance, the `extendsWith` function I showed would allow you to merge extensions however you want, changing the next layer (and therefore `final`) to have a wildly different structure, possibly.

---

Considering the current implementation, the most general input and output that won't cause type error would be

```suggestion
    toExtension ::
      b' -> Any -> Any -> b'

    or

    toExtension ::
      (a -> b') -> Any -> a -> b'

    or

    toExtension ::
      (a -> a -> b') -> a -> a -> b'

    where b' = ! Callable

    If a = b = b' = AttrSet & ! Callable, toExtension returns an extending function.
```